Adds `limit` parameter to `ansible.builtin.find`

colin-nolan opened the pull request:
##### SUMMARY

<!--- Describe the change below, including rationale and design decisions -->
`ansible.builtin.find` currently looks for matches in all non-filtered files.  When there is a very big file system in the search scope, and/or an expensive search is made (e.g. based off whole file contents), the operation can be very expensive. 

This PR adds the ability for the user to limit the number of matches that are made. Use cases that would benefit/be enabled with this functionality include:
- Determining if there are `> n` matching files in a directory.
- Finding the first directory in a nested tree with any files.
- Discovering where the file with matching content is (stopping once found!). 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Feature Pull Request

##### ADDITIONAL INFORMATION

<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

Try the module using: 
```bash
# Can't run module in place due to import issues caused by other modules in directory
cp lib/ansible/modules/find.py /tmp/find.py && python /tmp/find.py <<EOF
{
    "ANSIBLE_MODULE_ARGS": {
        "paths": "/var/log",
        "file_type": "file",
        "contains": "wally",
        "read_whole_file": true,
        "patterns": "^.*\\\.log$",
        "use_regex": true,
        "recurse": true,
        "limit": 1
    }
}
EOF
```

colin-nolan committed:
Allow find to be limited to max_matches.

colin-nolan committed:
Support max_matches when multiple in directory.

colin-nolan committed:
Simplifies max_matches validation.

colin-nolan committed:
Describes max_matches in docs.

colin-nolan committed:
Explicitly guarantee topdown search.

colin-nolan committed:
Adds example using max_matches.

colin-nolan committed:
Adds pedantic comma in doc.

colin-nolan committed:
Corrects version_added for max_matches.

colin-nolan commented:
I appreciate you likely have more important work to deal with but tagging module author @bcoca, as it feels like ansibot should have.

bcoca commented on lib/ansible/modules/find.py:
```diff
@@ -154,6 +154,15 @@
+            - Set to V(-1) for unlimited matches.
+            - Default is unlimited matches.
+        type: int
+        default: -1
```
use `None` for 'unset/unlimited'

colin-nolan replied:
Will change. Am I correct in thinking that the type should still be `int`, given there's no `Optional[int]` type (https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#argument-spec)?

colin-nolan replied:
Have changed. I'm not sure what the correct value to set as default in the docs should be now that it's `None` (see test failure). Any help would be appreciated.

bcoca replied:
you can just remove it, `None` is the default for `default`. All types allow `None` as a default as it is 'unset'.

colin-nolan replied:
Thanks - changed. Tests passing again.

bcoca commented on lib/ansible/modules/find.py:
```diff
@@ -154,6 +154,15 @@
             - When doing a C(contains) search, determine the encoding of the files to be searched.
         type: str
         version_added: "2.17"
+    max_matches:
```
i would rename `limit` or 'maximum` as this does not relate to a 'match' field, but to all the conditions the other options can set

colin-nolan replied:
Re-named to `limit`.

bcoca commented on lib/ansible/modules/find.py:
```diff
@@ -154,6 +154,15 @@
         version_added: "2.17"
+    max_matches:
+        description:
+            - Set the maximum number of matches to make before returning.
```
more like:
The maximum number of matching paths returned, after finding this many the find action will stop looking.

colin-nolan replied:
Updated

bcoca commented on lib/ansible/modules/find.py:
```diff
@@ -530,7 +553,8 @@ def handle_walk_errors(e):
 
-            for root, dirs, files in os.walk(npath, onerror=handle_walk_errors, followlinks=params['follow']):
+            # Setting `topdown=True` to explicitly guarantee matches are made from the shallowest directory first
+            for root, dirs, files in os.walk(npath, onerror=handle_walk_errors, followlinks=params['follow'], topdown=True):
```
this changes the normal search by adding topdown, this is not really backwards compatible

bcoca replied:
i would even add this as an option before i limit matching

colin-nolan replied:
`topdown` is `True` by default (https://docs.python.org/3/library/os.html#os.walk) so it should not change the current behaviour. The intention was to make it explicit that the module is guaranteed to be topdown - perhaps the docs is a better place for this though?

I'm not sure how many people will have a use-case where they'd want to switch between topdown/bottomup. However, I'm happy to add it if you think it could be useful.

bcoca reviewed (changes requested):
I'm not even sure this feature is needed as it is a simple thing to cut down the return of files to only those you need `returned_files[:5]`. 

Though this can be a performance improvement, it would only be really noticeable with huge amounts of files.

colin-nolan committed:
Makes max matches unlimited == None.

colin-nolan committed:
Rename max_matches -> limit.

colin-nolan committed:
Improves description of limit param.

colin-nolan committed:
Corrects logic for invalid limit.

colin-nolan committed:
Improve msg returned when limit reached.

colin-nolan committed:
Don't specify `None` default.

colin-nolan commented:
@bcoca many thanks for reviewing. I have addressed most of your comments. Just two things left to resolve:
1. How to document a default of `None`, where `None` is not of type `int` (currently failing a check).
2. Whether you think an option to change between topdown/bottom up is required. I'd personally suggest putting it out of scope, as it's difficult to think of many use-cases where it would be useful.

> I'm not even sure this feature is needed as it is a simple thing to cut down the return of files to only those you need returned_files[:5].
>
> Though this can be a performance improvement, it would only be really noticeable with huge amounts of files.

This is the exact situation that I have - lot's of files, some of them very large. The module isn't really viable for my use-case without this option.

colin-nolan committed:
Remove default from limit description.

Implies default, which is none.

bcoca commented on lib/ansible/modules/find.py:
```diff
@@ -154,6 +154,14 @@
+        description:
+            - Limit the maximum number of matching paths returned. After finding this many, the find action will stop looking.
+            - Matches are made from the top, down (i.e. shallowest directory first).
+            - Set to V(None) for unlimited matches.
```
`If not set, or set to v(null), it will do unlimited matches`

colin-nolan replied:
```suggestion
            - If not set, or set to v(null), it will do unlimited matches
```

colin-nolan replied:
Changed

colin-nolan committed:
Improve limit doc for when not set/none

colin-nolan committed:
Pedantic full stop in doc.
